src/transaction/log_recovery.c

```diff
@@ -491,6 +511,356 @@ log_rv_redo_record (THREAD_ENTRY * thread_p, log_reader & log_pgptr_reader,
+		     (long long int) log_rec.undo_nxlsa.pageid, (int) log_rec.undo_nxlsa.offset);
+    }
+}
+
+/*
+ * log_rv_redo_record_sync_or_dispatch_parallel - execute a redo record synchronously or
+ *                    (in future) asynchronously
+ *
+ * return: nothing
+ *
+ *   thread_p(in):
+ *   log_pgptr_reader(in/out): log reader
+ *   log_rec(in): log record structure with info about the location and size of the data in the log page
+ *   rcv_lsa(in): Reset data page (rcv->pgptr) to this LSA
+ *   log_rtype(in): log record type needed to check if diff information is needed or should be skipped
+ *   undo_unzip_support(out): extracted undo data support structure (set as a side effect)
+ *   redo_unzip_support(out): extracted redo data support structure (set as a side effect); required to
+ *                    be passed by address because it also functions as an internal working buffer;
+ *                    functions as an outside managed (ie: longer lived) buffer space for the underlying
+ *                    logic to perform its work; the pointer to the buffer is then passed - non-owningly
+ *                    - to the rcv structure which, in turn, is passed to the actual apply function
+ *
+ */
+// *INDENT-OFF*
+template <typename T>
+void log_rv_redo_record_sync_or_dispatch_parallel (THREAD_ENTRY * thread_p, log_reader & log_pgptr_reader,
+                                                   const T & log_rec, const log_lsa & rcv_lsa,
+                                                   const LOG_LSA * end_redo_lsa, LOG_RECTYPE log_rtype,
+                                                   LOG_ZIP & undo_unzip_support, LOG_ZIP & redo_unzip_support)
+// *INDENT-ON*
+{
+  // *INDENT-OFF*
+  const VPID rcv_vpid = log_rv_get_log_rec_vpid<T> (log_rec);
+  // *INDENT-ON*
+  // at this point, vpid can either be valid or not
+
+  // TODO: once valid vpid is extracted, it can be decided whether to execute sync or dispatch asynchronously
+  // as such, from here on, everything should be moved into another function
+
+  // *INDENT-OFF*
+  const LOG_DATA & log_data = log_rv_get_log_rec_data<T> (log_rec);
+  // *INDENT-ON*
+
+  LOG_RCV rcv;
+  if (!log_rv_fix_page_and_check_redo_is_needed (thread_p, rcv_vpid, rcv, log_data.rcvindex, rcv_lsa, end_redo_lsa))
+    {
+      /* nothing else needs to be done, see explanation in function */
+      assert (rcv.pgptr == nullptr);
+      return;
+    }
+  // at this point, pgptr can be null or not
+
+  /* will take care of unfixing the page, will be correctly de-allocated as it is the same
+   * storage class as 'rcv' and allocated on the stack after 'rcv' */
+  log_rcv_raii_pgptr raii_pgptr (thread_p, rcv);
+
+  // *INDENT-OFF*
+  rcv.length = log_rv_get_log_rec_redo_length<T> (log_rec);
+  rcv.mvcc_id = log_rv_get_log_rec_mvccid<T> (log_rec);
+  rcv.offset = log_rv_get_log_rec_offset<T> (log_rec);
+
+  log_rv_redo_record_debug_logging_pre_data_retrieve<T> (rcv_lsa, log_data.rcvindex, rcv_vpid, rcv);
```

Comment:
It would work, but it also requires to take an explicit parameter which - as it stands now - is an implementation detail of the function. Also, a `const char*` param would allow a "lazy" user to just pass in gibberish.
I prefer it as a template function.

Reply:
alright